File: tests/workflow/producer/rust/src/producer/responses/structc.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use uuid::Uuid;

pub enum Response {
    CaseB(protocol::StructB),
    CaseF(protocol::StructF),
    CaseD(protocol::StructD),
}

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::StructC,
    scope: &mut Scope<'_, E, C>,
) -> Result<Response, protocol::StructE> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::StructC);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructB);
        Ok(Response::CaseB(samples::struct_b::get()))
    } else if index == 2 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructF);
        Ok(Response::CaseF(samples::struct_f::get()))
    } else if index == 3 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructD);
        Ok(Response::CaseD(samples::struct_d::get()))
    } else {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructE);
        Err(samples::struct_e::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/structc.ts
import { Response } from "../implementation/responses/structc";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.StructC,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.StructC);
	const index = stat.index(Alias.StructC);
	if (index === 0) {
		return Promise.resolve(new Response(Protocol.StructB.defaults()));
	} else if (index === 1) {
		return Promise.resolve(new Response(Protocol.StructF.defaults()));
	} else if (index === 2) {
		return Promise.resolve(new Response(Protocol.StructD.defaults()));
	} else if (index === 3) {
		stat.case(Alias.StructE);
		return Promise.resolve(new Response(Protocol.StructE.defaults()));
	} else {
		panic(`Unexpected index for Protocol.StructC request: ${index}`);
		return Promise.reject(
			new Error(`Handler for Protocol.StructC isn't implemented.`)
		);
	}
}
